lib/cdo/aws/cloud_formation.rb

```diff
@@ -0,0 +1,158 @@
+    INSTANCE_TYPE = 'c4.xlarge'
+    SSH_IP = '0.0.0.0/0'
+
+    # Use AWS Certificate Manager for ELB and CloudFront SSL certificates.
+    CERTIFICATE_ARN = Aws::ACM::Client.new(region: 'us-east-1').
+      list_certificates(certificate_statuses: ['ISSUED']).
+      certificate_summary_list.
+      find{|cert| cert.domain_name == "*.#{DOMAIN}"}.
+      certificate_arn
+
+    def self.cfn
+      @@cfn ||= Aws::CloudFormation::Client.new
+    end
+
+    def self.update_certs
+      Dir.chdir(aws_dir('cloudformation')) do
+        RakeUtils.bundle_exec './update_certs', FQDN
+      end
+    end
+
+    def self.update_cookbooks
+      if CDO.chef_local_mode
+        RakeUtils.with_bundle_dir(cookbooks_dir) do
+          Tempfile.open('berks') do |tmp|
+            RakeUtils.bundle_exec 'berks', 'package', tmp.path
+            Aws::S3::Client.new(region: 'us-east-1').put_object(
+              bucket: 'cdo-dist',
+              key: "chef/#{RakeUtils.git_branch}.tar.gz",
+              body: tmp.read
+            )
+          end
+        end
+      end
+    end
+
+    def self.update_bootstrap_script
+      Aws::S3::Client.new(region: 'us-east-1').put_object(
+        bucket: 'cdo-dist',
+        key: 'chef/bootstrap.sh',
+        body: File.read(aws_dir('chef-bootstrap.sh'))
+      )
+    end
+
+    def self.wait_for_stack(action, start_time)
+      CDO.log.info "Stack #{action} requested, waiting for instance provisioning to complete..."
+      begin
+        cfn.wait_until("stack_#{action}_complete".to_sym, stack_name: STACK_NAME) do |w|
+          w.max_attempts = 360 # 1 hour
+          w.delay = 10
+          w.before_wait { print '.' }
+        end
+      rescue Aws::Waiters::Errors::FailureStateError
+        CDO.log.info "\nError on #{action}. Event log:"
+        events = []
+        cfn.describe_stack_events(stack_name: STACK_NAME).stack_events.each do |event|
+          events << event
+          break if event.timestamp < start_time
+        end
+        events.reject{|event|event.resource_status_reason.nil?}.sort_by(&:timestamp).each do |event|
+          puts "#{event.timestamp}- #{event.logical_resource_id} [#{event.resource_status}]: #{event.resource_status_reason}"
+        end
+        return
+      end
+      CDO.log.info "\nStack #{action} complete."
+    end
+
+    def self.delete
+      stack_exists = !!cfn.describe_stacks(stack_name: STACK_NAME) rescue false
+      if stack_exists
+        CDO.log.info "Shutting down #{STACK_NAME}..."
+        start_time = Time.now
+        cfn.delete_stack(stack_name: STACK_NAME)
+        wait_for_stack(:delete, start_time)
+      else
+        CDO.log.warn "Stack #{STACK_NAME} does not exist."
+      end
+    end
+
+    def self.create_or_update(cdn_enabled=false)
+      json_template = json_template(cdn_enabled)
+
+      update_certs
+      update_cookbooks
+      update_bootstrap_script
+
+      stack_exists = !!cfn.describe_stacks(stack_name: STACK_NAME) rescue false
```

Comment:
rescue false is dangerous because it will mask bugs in the calling code and swallow transient errors in describe_stacks.
 
is it possible to be more specific in what you are rescuing?

Do we need to retry on transient failure here or does the client library do that for you?